openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -74,19 +77,37 @@ func NewCinderProvisioner(client kubernetes.Interface, id, configFilePath string
 // Provision creates a storage asset and returns a PV object representing it.
 func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
 	var (
-		connection volumeservice.VolumeConnection
-		mapper     volumeMapper
-		pv         *v1.PersistentVolume
-		cleanupErr error
+		connection  volumeservice.VolumeConnection
+		mapper      volumeMapper
+		pv          *v1.PersistentVolume
+		volumeID    string
+		err         error
+		cleanupErr  error
+		found       bool
+		volumeState string
 	)
 
 	if options.PVC.Spec.Selector != nil {
 		return nil, fmt.Errorf("claim Selector is not supported")
 	}
 
+	if volumeID, found = options.PVC.Annotations[ShouldProvisionVolumeAnn]; found {
+		volumeState, err = p.vsb.getCinderVolumeState(p.VolumeService, volumeID)
+		if err != nil {
+			glog.Errorf("Failed to get volume %s, err: %v", volumeID, err)
+			goto ERROR
+		}
+		if volumeState != "available" {
+			glog.Errorf("Volume %s state is not available, current state %v", volumeID, volumeState)
+			err = fmt.Errorf("Volume %s state is not available, current state %v", volumeID, volumeState)
+			goto ERROR
+		}
+		goto CONNECT
```

Comment:
We should record the state that this is a pre-existing volume so if we encounter an error later (eg. connecting or reserving) we won't delete the volume.

Reply:
yeah, I'm also thinking of this, I'll add it